usecases/sharding/remote_index.go

```diff
@@ -504,6 +506,11 @@ func (ri *RemoteIndex) queryReplicas(
 			if resp, err = queryOne(node); err == nil {
 				return resp, node, nil
 			}
+			ri.log.WithFields(logrus.Fields{
```

Comment:
will this capture the `ctx.Err()` early return (happening on this same loop before your log)? Meaning we may miss if retry failed because `context.DeadlineExceeded` or `context.Timeout`?

Few clarifications:
1. Since we are logging inside the loop, may I know what kind of timeout we have for each retry? 
2. Also why it's Debug? used only for debugging (debug), actionable (error, warn), just a info (info)?

I'm trying to correlate this log line with ones on `retrier` above. So we log every query replicas(say `m`) and each replica can be queried with `n` retries?. 

Am I understanding this right? 
Worst case if all the retry failed we have `mxn` log lines to make sense of what happened? To be that looks noisy. 

Instead of logging these many times can we just collect info and log once outside the loop? Of course this may not work if each `queryReplicas` and `eachRetry` have bigger timeout and backoff, say I have to wait so long to see the final log line in order to understand what's going on. That's my first question what kind of timeout and backoff delay we talking about?

Reply:
Thanks, these are great thoughts/questions.

> will this capture the ctx.Err() early return (happening on this same loop before your log)? Meaning we may miss if retry failed because context.DeadlineExceeded or context.Timeout?

If we hit the `ctx.Err` early return, then we would not log anything and would return the error and not retry anymore. I think ensuring there is a log from one of the callers higher up in the call stack for any error returned from `queryReplicas` would be helpful (I would guess right now the error is returned to the user but I'm not sure if it's logged, will check). My general principle in Go is: Callers are responsible for logging returned errors while callees are responsible for providing useful messages in the returned errors and callees are responsible for logging any "useful info" not captured in their return values.

> Since we are logging inside the loop, may I know what kind of timeout we have for each retry?

The current implementation tends to have anywhere from 9-34 retries, with jittered exponential backoff starting at 250ms and maxing out at 30s per retry (see `adapters/clients/client.go` for more details).

> Also why it's Debug? used only for debugging (debug), actionable (error, warn), just a info (info)?

I agree that it doesn't make sense as debug, but am worried about spamming the logs, I like your idea of doing something like collecting info and logging once so that we could make this an info or warn log.

> Worst case if all the retry failed we have mxn log lines to make sense of what happened? To be that looks noisy.

Correct, agreed that it's too noisy.

> Instead of logging these many times can we just collect info and log once outside the loop?

I'm thinking through this now, if we log every time trying a given replica has failed all it's retries, (assuming the context never expires) we're looking at a delay of `9 retries * 15s average delay = 135s` (simplifying math with 15s average delay assumption) to `34 retries * 15s average delay = 510s`. Both of those are long enough to make me hesitant but it's hard to say whether it's worse to have spammy logs vs waiting 2-10 minutes before you can tell what is happening during retries. If the context expires before the retries are done, then you wouldn't get any information about the retries with the current code, which is quite bad.

I'm going to give this more thought, I'm happy you brought up the spaminess and shared some ideas for improving it.